### PR TITLE
ci: Add Cloud Hypervisor / QEMU jobs to test with the "devmapper" snapshotter

### PR DESCRIPTION
**Pull request by fidencio**
ci: Add USE_DEVMAPPER flag

Currently selecting whether devmapper should (or should not) be used as
the containerd snapshotter is done based on the HYPERVISOR used with
Kata Containers, leading the only testing it when using Firecracker.

In order to prepare the ground for adding new jobs, let's ensure we
un-tie its configuration from the hypervisor being used.

---

ci: Add cloud-hypervisor + devmapper CI

In order to ensure situations like the ones reported on the following
issues* won't happen again, let's add a new CI job covering Cloud
Hypervisor, running with containerd as the CRI Engine, having devmapper
as its snapshotter.

---

ci: Add QEMU + devmapper CI

In order to ensure situations like the ones reported on the following
issues* won't happen again, let's add a new CI job covering QEMU,
running with containerd as the CRI Engine, having devmapper as its
snapshotter.

---

*:
https://github.com/kata-containers/kata-containers/issues/4005
https://github.com/kata-containers/kata-containers/issues/4043

Fixes: #4688

### CODE REVIEW

**Comment by fidencio** [comment]
Note, we still do **not** have any job actually using this.  Those will be created shortly and tested in this PR, preferably before this one gets merged.

With that being said, we want to make sure we don't break anything with this new addition, so /test

**Comment by fidencio** [comment]
We also need
```
diff --git a/.ci/containerd_devmapper_setup.sh b/.ci/containerd_devmapper_setup.sh
index 7dce9e4e..ef75acbf 100755
--- a/.ci/containerd_devmapper_setup.sh
+++ b/.ci/containerd_devmapper_setup.sh
@@ -12,13 +12,13 @@ set -o pipefail
 cidir=$(dirname "$0")
 source "${cidir}/../lib/common.bash"

-KATA_HYPERVISOR="${KATA_HYPERVISOR:-qemu}"
+USE_DEVMAPPER="${USE_DEVMAPPER:-false}"

 containerd_config_dir="/etc/containerd"
 containerd_config_file="${containerd_config_dir}/config.toml"

-if [ "$KATA_HYPERVISOR" != "firecracker" ]; then
-       echo "WARNING: Devicemapper configuration is only for Firecracker. Exiting"
+if [ "$USE_DEVMAPPER" != "true" ]; then
+       echo "WARNING: Devicemapper configuration is not configured. Exiting"
        exit 0
 fi
```

Otherwise devmapper won't be configured.  I'm squashing that info the very first patch of the series.

**Comment by fidencio** [comment]
And another one to be squashed into the first one:
```
diff --git a/.ci/setup.sh b/.ci/setup.sh
index a60973b2..191dec43 100755
--- a/.ci/setup.sh
+++ b/.ci/setup.sh
@@ -125,7 +125,7 @@ install_extra_tools() {
        fi

        if [ "${USE_DEVMAPPER}" == "true" ]; then
-               info "Configure devicemapper for firecracker"
+               info "Configure devicemapper"
                bash -f "${cidir}/containerd_devmapper_setup.sh"
        fi

```

I will let CI finish though before pushing a new series.

**Comment by fidencio** [comment]
Cloud Hypervisor failed and you can see in the container-shim-kata-v2 logs:
```
Apr 12 18:14:37 ubuntu1804-azurec82cc0 kata[4906]: time="2022-04-12T18:14:37.392957976Z" level=warning msg="stop container failed" container=20314b5e364ec5c365914e9bb9668653506a21ae4c8e64e442c81285333d1315 error="rpc error: code = Internal desc = failed to remove dir \"/run/kata-containers/shared/containers/20314b5e364ec5c365914e9bb9668653506a21ae4c8e64e442c81285333d1315\"\n\nCaused by:\n    Read-only file system (os error 30)" name=containerd-shim-v2 pid=4906 sandbox=702ea4aa53c6bcbb132e91012ec0ccaeb35c79b1d9b019504400233490fb485b source=containerd-kata-shim-v2
```

So, yeah, the issue is reproducible with this CI, and this CI would be enough to avoid the regression.
Let me update the series one last time to include https://github.com/kata-containers/tests/pull/4689#issuecomment-1097035714

**Review by egernst — APPROVED** [review]
Thanks @fidencio

**Review by gkurz — COMMENTED** [review]
What about this snippet from `integration/containerd/cri/integration-tests.sh` ?

```
if [ "$KATA_HYPERVISOR" == "firecracker" ]; then
	sudo sed -i 's|^\(\[plugins\]\).*|\1\n  \[plugins.devmapper\]\n    pool_name = \"contd-thin-pool\"\n    base_image_size = \"4096MB\"|' ${CONTAINERD_CONFIG_FILE}
	echo "Devicemapper configured"
	cat "${CONTAINERD_CONFIG_FILE}"
fi
```

This looks like a good candidate for the `Add USE_DEVMAPPER flag` patch.

Apart from that, the rest LGTM.

**Comment by fidencio** [comment]
@gkurz, updated, thanks for the review.

File: .ci/ci_job_flags.sh [hunk 1]
```diff
@@ -142,6 +142,14 @@ case "${CI_JOB}" in
 	export KATA_HYPERVISOR="cloud-hypervisor"
 	export KUBERNETES="yes"
 	;;
+"CLOUD-HYPERVISOR-K8S-CONTAINERD-DEVMAPPER")
```
Comment (c3d) [inline .ci/ci_job_flags.sh:153]: I believe that the other names use underscore `_` and not hyphen `-`.

File: .ci/run.sh [hunk 2]
```diff
@@ -81,7 +81,7 @@ case "${CI_JOB}" in
 		echo "INFO: Running kubernetes tests"
 		sudo -E PATH="$PATH" bash -c "make kubernetes"
 		;;
-	"CLOUD-HYPERVISOR-K8S-CONTAINERD")
+	"CLOUD-HYPERVISOR-K8S-CONTAINERD"|"CLOUD-HYPERVISOR-K8S-CONTAINERD-DEVMAPPER")
```
Comment (c3d) [inline .ci/run.sh:84]: Ah… well, it looks like this was not entirely consistent throughout.

**Comment by c3d** [comment]
Thanks @fidencio.

**Review by c3d — APPROVED** [review]
Looks like a step in the right direction, @fidencio. Thanks.

**Review by gkurz — APPROVED** [review]
Thanks @fidencio !

**Comment by gkurz** [comment]
The failing jobs are hitting #4692, being worked on by @liubin.

**Review by egernst — APPROVED** [review]

**Comment by fidencio** [comment]
I like Eric's double approval. :-)

**Review by egernst — APPROVED** [review]
A still approve!

**Pull request merged**